dhall/src/Dhall/Pretty/Internal.hs

```diff
@@ -1072,14 +1072,37 @@ prettyCharacterSet characterSet expression =
         angles . map prettyAlternative . Dhall.Map.toList
 
     prettyChunks :: Pretty a => Chunks Src a -> Doc Ann
-    prettyChunks (Chunks a b) =
-        if anyText (== '\n')
-        then
+    prettyChunks (Chunks a b)
+        | anyText (\c -> Data.Char.isControl c && not (Data.Char.isSpace c)) =
+            short
+        | leadingWhitespaceBeforeClosing'' =
+            short
+        | anyText (== '\n') =
             if anyText (/= '\n')
             then long
             else Pretty.flatAlt long short
-        else short
+        | otherwise =
+            short
       where
+        anyText predicate = any (\(text, _) -> Text.any predicate text) a || Text.any predicate b
+
+        -- If we format the text literal as a multi-line string, would the last
+        -- line start with whitespace? (Leading whitespace on the line of the
+        -- closing @''@ is ignored by the parser, see #1545.)
+        leadingWhitespaceBeforeClosing'' =
+            case lineStarts of
+                []                -> False
+                ts | t <- last ts -> " " `Text.isPrefixOf` t || "\t" `Text.isPrefixOf` t
+          where
+            lineStarts =
+                    concatMap (Text.splitOn "\n") (take 1 texts)
+                ++  concatMap
+                        -- Each first line of these text chunks is preceded by a
+                        -- string interpolation, so we remove these with (drop 1).
+                        (drop 1 . (Text.splitOn "\n"))
+                        (drop 1 texts)
+            texts = map fst a ++ [b]
```

Comment:
This is more complex and probably less efficient than I'd desire. Maybe I can still come up with something better…

Reply:
Yeah, we should probably scrutinize `b` before looking at the other text chunks.